File: solutions/42/main.py
from typing import List 

class Solution:
    def trap(self, height: List[int]) -> int:
        result = 0
        st = []
        st.append(0)
        for i in range(1, len(height)):
            if height[st[-1]] > height[i]:
                st.append(i)
            elif height[st[-1]] == height[i]:
                st[-1] = i
            else:
                while len(st) > 0 and height[st[-1]] < height[i]:
                    mid = height[st[-1]]
                    st.pop()
                    if len(st) > 0:
                        w = i - st[-1] - 1
                        h = min(height[i], height[st[-1]])
                        result += w * (h - mid)
            st.append(i)
        return result

if __name__ == "__main__": 
    height = [0,1,0,2,1,0,1,3,2,1,2,1]
    s = Solution()
    assert s.trap(height) == 6

File: solutions/42/main.cpp
#include <iostream>
#include <vector>
#include <gtest/gtest.h>

using namespace std; 

class Solution {
public:
    int trap(vector<int>& height) {

        const int n = height.size(); 
        vector<int> max_left(n);
        vector<int> max_right(n);  

        for (int i = 1; i < n; ++i)
        {
            // find the max height on the left 
            max_left[i] = max(max_left[i - 1], height[i - 1]);
            // find the max height on the right 
            max_right[n - 1 - i] = max(max_right[n - i], height[n - i]); 
        }

        int sum = 0; 
        for (int i = 0; i < n; ++i)
        {
            int min_height = min(max_left[i], max_right[i]);
            if (min_height > height[i])
            {
                sum += min_height - height[i]; 
            }
        }

        return sum; 
    }
};

TEST(Test42, SimpleTest) 
{
    vector<int> height = {0,1,0,2,1,0,1,3,2,1,2,1}; 
    Solution s; 
    EXPECT_EQ(s.trap(height), 6) << "Wrong answer"; 
}

int main() 
{
    testing::InitGoogleTest(); 
    return RUN_ALL_TESTS();
}
